_posts/2022-04-26-type-based-optimizations-in-the-jit.md

```diff
@@ -0,0 +1,804 @@
+    {test,is_tuple,{f,3},[{x,0}]}.
+    {get_tuple_element,{x,0},1,{x,0}}.
+    {deallocate,1}.
+    return.
+  {label,3}.
+    {move,{y,0},{x,0}}.
+    {deallocate,1}.
+    return.
+```
+
+The compiler has figured out that the `increment/1` returns either the
+atom `error` or a two-tuple with `ok` as the first element. Therefore,
+to distinguish between those two possible return values, a single
+instruction suffices:
+
+```
+    {test,is_tuple,{f,3},[{x,0}]}.
+```
+
+There is no need to explicitly test for the value `error` because it
+**must** be `error` if it is not a tuple. Similarly, there is no need
+to test that the first element of the tuple is `ok` because it must be.
+
+In OTP 24, the JIT translates that instruction to a sequence of 5 native
+instructions for x86_64:
+
+```
+# i_is_tuple_fs
+    mov rsi, qword ptr [rbx]
+    rex test sil, 1
+    jne L2
+    test byte ptr [rsi-2], 63
+    jne L2
+```
+
+(Lines starting with `#` are comments.)
+
+The `mov` instruction fetches the value of the BEAM register `{x,0}`
+to the CPU register `rsi`. The next two instructions test whether the
+term is a pointer to an object on the heap. If it is, the header word
+for the heap object is tested to make sure it is a tuple. The second
+test is needed because the heap object could be some other Erlang term,
+such as a binary, a map, or an integer that does not fit in a machine
+word.
+
+Now let's see what the compiler and the JIT in OTP 25 do with this
+instruction. The BEAM code looks like this:
+
+```
+    {test,is_tuple,
+          {f,3},
+          [{tr,{x,0},
+               {t_union,{t_atom,[error]},
+                        none,none,
+                        [{{2,{t_atom,[ok]}},
+                          {t_tuple,2,true,
+                                   #{1 => {t_atom,[ok]},
+                                     2 => {t_integer,any}}}}],
+                        none}}]}.
+```
+
+The operand that was `{x,0}` in OTP 24 is now a tuple:
+
+```
+{tr,Register,Type}
+```
+
+That is, it is a three-tuple with `tr` as the first element. `tr`
+stands for **typed register**. The second element is the BEAM register
+(`{x,0}` in this case), and the third element is the type of the
+register in the compiler's internal type representation. The type
+is equivalent to the following type spec:
+
+```
+'error' | {'ok', integer()}
+```
+
+The JIT has (currently) no need for that level of details in the
+types, so the compiler embeds a [simplified
+version](https://github.com/erlang/otp/blob/de5bb49320db22159de52e677c5f7499b763b0cd/lib/compiler/src/beam_types.erl#L1153-L1241)
+of that type into the BEAM file. The embedded type is equivalent to:
+
+```
+atom() | tuple()
+```
+
+By knowing that `{x,0}` must be an atom or a tuple, the JIT in OTP 25
+emits the following simplified native code:
+
+```
+# i_is_tuple_fs
+    mov rsi, qword ptr [rbx]
+# simplified tuple test since the source is always a tuple when boxed
+    rex test sil, 1
+    jne label_3
+```
```

Comment:
Just a general thought that maybe it would be more illustrative if the generated native code for OTP 24 and OTP 25 could be presented side by side or very close to each other so that the difference is easy to see. Maybe it is ok to start with the 2 variants of generated code and then explain how and why they differ. Or if it is ok to show another copy of the OTP 24 code close to the OTP 25 variant. Just a thought , the explaining text is really good.

Reply:
I am not sure how to do that in Markdown. It could possibly be done with some table extensions, but I think that it could get too wide and users would have to scroll.